apps/hash-graph/lib/authorization/src/backend/spicedb.rs

```diff
@@ -0,0 +1,303 @@
+                serialize.serialize_field("optionalRelation", &relation)?;
+                serialize.end()
+            } else {
+                let mut serialize = serializer.serialize_struct("SubjectReference", 1)?;
+                serialize.serialize_field("object", &ResourceReference(object))?;
+                serialize.end()
+            }
+        }
+    }
+
+    /// Represents a reference to a caveat to be used by caveated relationships.
+    #[derive(Debug, Serialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct ContextualizedCaveat<'a> {
+        /// The name of the caveat expression to use, as defined in the schema.
+        pub caveat_name: &'a str,
+        /// Consists of key-value pairs that will be injected at evaluation time.
+        ///
+        /// The keys must match the arguments defined on the caveat in the schema.
+        pub context: HashMap<&'a str, serde_json::Value>,
+    }
+
+    pub struct RelationshipTuple<R, A, S>(pub Tuple<R, A, S>);
+    impl<R: Resource, A: Affiliation<R>, S: Subject> Serialize for RelationshipTuple<R, A, S> {
+        fn serialize<Ser: Serializer>(&self, serializer: Ser) -> Result<Ser::Ok, Ser::Error> {
+            let mut serialize = serializer.serialize_struct("Tuple", 3)?;
+            serialize.serialize_field("resource", &ResourceReference(&self.0.resource))?;
+            serialize.serialize_field("relation", &self.0.affiliation)?;
+            serialize.serialize_field("subject", &SubjectReference(&self.0.subject))?;
+            serialize.end()
+        }
+    }
+
+    /// Specifies how a resource relates to a subject.
+    ///
+    /// Relationships form the data for the graph over which all permissions questions are answered.
+    #[derive(Serialize)]
+    #[serde(
+        rename_all = "camelCase",
+        bound = "R: Resource, A: Affiliation<R>, S: Subject"
+    )]
+    pub struct Relationship<'a, R, A, S> {
+        #[serde(flatten)]
+        pub tuple: RelationshipTuple<R, A, S>,
+        #[serde(skip_serializing_if = "Option::is_none", rename = "optionalCaveat")]
+        pub caveat: Option<ContextualizedCaveat<'a>>,
+    }
+}
+
+#[derive(Debug)]
+pub struct SpiceDb {
+    pub configuration: SpiceDbConfig,
+}
+
+#[derive(Debug)]
+enum InvocationError {
+    Request(reqwest::Error),
+    Rpc(RpcStatus),
+}
+
+impl fmt::Display for InvocationError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Request(error) => fmt::Display::fmt(error, fmt),
+            Self::Rpc(status) => write!(fmt, "Error {}: {}", status.code, status.message),
+        }
+    }
+}
+
+impl Error for InvocationError {}
+
+enum InvocationResult<T> {
+    Success(T),
+    Failure(RpcStatus),
+}
+
+impl SpiceDb {
+    async fn request<R: DeserializeOwned>(
+        &self,
+        path: &'static str,
+        body: &(impl Serialize + Sync),
+    ) -> Result<InvocationResult<R>, reqwest::Error> {
```

Comment:
I do not particularly like that we essentially return `Result<Result<R, RpcStatus>, request::Error>` here, IMO only a single `Result` should be returned, e.g.

```rust
enum RequestError {
	Request(reqwest::Error),
	Remote(RpcStatus)
}

Result<R, RequestError>
```

Reply:
I mean `InvocationError` already exists.

Reply:
Changed in [`0a2ba99` (#3014)](https://github.com/hashintel/hash/pull/3014/commits/0a2ba99b9a56408bb5bceede1bf323f27e759310)